pkg/kv/kvserver/snapshot_apply_prepare.go

```diff
@@ -51,15 +55,12 @@ func prepareSnapApply(
 		return roachpb.Span{}, nil, err
 	}
 
-	if len(input.subsumedDescs) > 0 {
-		spans, err := clearSubsumedReplicaDiskData(
-			ctx, input.st, input.todoEng, input.writeSST,
-			input.desc, input.subsumedDescs,
-		)
-		if err != nil {
-			return roachpb.Span{}, nil, err
-		}
-		clearedSubsumedSpans = spans
+	clearedSubsumedSpans, err := clearSubsumedReplicaDiskData(
```

Comment:
I timed out on a couple of minutes looking at `clearSubsumedReplicaDiskData` trying to understand why it's a no-op if `subsumedDescs` is empty. It probably indicates that it's better to make it explicit (in its comment and/or have an `if` in its body).

Particularly, why is this second loop in it a no-op?
https://github.com/cockroachdb/cockroach/blob/df7d3153243b3b680cdc2c7b8474bbd6f11ed9e9/pkg/kv/kvserver/replica_raftstorage.go#L930

We used to call this unconditionally though (so the change is good), is the `if len > 0` a recent addition? https://github.com/cockroachdb/cockroach/blob/df7d3153243b3b680cdc2c7b8474bbd6f11ed9e9/pkg/kv/kvserver/replica_raftstorage.go#L553-L564.

Reply:
On a second attempt to read it, I see why it's no-op: it compares keys and comparisons never succeed. I still think we should make the zero case explicit to remove doubt, and also seems like there is wasted work there generating and comparing keys.

Reply:
> is the if len > 0 a recent addition?

Yes, this is unfortunate AI jank - it was introduced in "kvserver: move subsumed replica disk data clearing into prepareSnapshot" and I didn't catch it during self-review.

I'll add a comment about the case in which subsumedDescs is nil. I'm mildly -1 on bailing early in this method: if we don't trust this method to do the right thing if nothing got subsumed, how can we believe that it does the right thing in general?

Reply:
> how can we believe that it does the right thing in general?

By testing it :)

The argument is not in the lack of trust in the code, but it's in readability (which is only a proxy for trust). It took me maybe 5 minutes to figure out how the empty case behaves and spot that it's a no-op.

Reply:
LGTM, thanks. Mind popping this somewhere (e.g. in the `clearSubsumedReplicaDiskData` comment)?

```diff
diff --git a/pkg/kv/kvserver/replica_raftstorage.go b/pkg/kv/kvserver/replica_raftstorage.go
index e19716a5995..f1447d06094 100644
--- a/pkg/kv/kvserver/replica_raftstorage.go
+++ b/pkg/kv/kvserver/replica_raftstorage.go
@@ -849,6 +849,8 @@ func rewriteRaftState(
 // method requires that each of the subsumed replicas raftMu is held, and that
 // the Reader reflects the latest I/O each of the subsumed replicas has done
 // (i.e. Reader was instantiated after all raftMu were acquired).
+//
+// NB: does nothing if subsumedDescs is empty.
 func clearSubsumedReplicaDiskData(
        ctx context.Context,
        st *cluster.Settings,
```